packages/liveblocks-core/src/lib/IncrementalJsonParser.ts

```diff
@@ -0,0 +1,259 @@
+
+// Characters that can end partial keywords: n, u, l, t, r, e, f, a, s
+const NULL_KEYWORD_CHARS = new Set("null");
+const TRUE_KEYWORD_CHARS = new Set("true");
+const FALSE_KEYWORD_CHARS = new Set("false");
+
+/**
+ * Strips the last character from `str` if it is one of the chars in the given
+ * `chars` string.
+ */
+function stripChar(str: string, chars: string): string {
+  const lastChar = str[str.length - 1];
+  if (chars.includes(lastChar)) {
+    return str.slice(0, -1);
+  }
+  return str;
+}
+
+export class IncrementalJsonParser {
+  // Input
+  #sourceText: string = "";
+
+  // Output
+  #cachedJson?: JsonObject;
+
+  /** How much we've already parsed */
+  #scanIndex: number = 0;
+  /** Whether the last char processed was a backslash */
+  #escaped: boolean = false;
+  /**
+   * Start position of the last unterminated string, -1 if we're not inside
+   * a string currently.
+   *
+   * Example: '{"a": "foo'
+   *                 ^
+   */
+  #lastUnterminatedString: number = -1;
+  /**
+   * Start position of the last fully terminated string we've seen.
+   *
+   * Example: '{"a": "foo'
+   *            ^
+   */
+  #lastTerminatedString: number = -1;
+  /** The bracket stack of expected closing chars. For input '{"a": ["foo', the stack would be ['}', ']']. */
+  #stack: string[] = [];
+
+  constructor(text: string = "") {
+    this.append(text);
+  }
+
+  get source(): string {
+    return this.#sourceText;
+  }
+
+  get json(): JsonObject {
+    if (this.#cachedJson === undefined) {
+      this.#cachedJson = this.#parse();
+    }
+    return this.#cachedJson;
+  }
+
+  /** Whether we're currently inside an unterminated string, e.g. '{"hello' */
+  get #inString(): boolean {
+    return this.#lastUnterminatedString >= 0;
+  }
+
+  append(delta: string): void {
+    if (delta) {
+      // Trim leading whitespace only on the first delta
+      if (this.#sourceText === "") {
+        delta = delta.trimStart();
+      }
+      this.#sourceText += delta;
+      this.#cachedJson = undefined; // Invalidate the cache
+    }
+  }
+
+  #autocompleteKeywords(output: string): string {
+    // Complete unambiguous partial JSON keywords,
+    // e.g. '{"a": -' → '{"a": -0'
+    //      '{"a": n' → '{"a": null'
+    //      '{"a": t' → '{"a": true'
+    //      '{"a": f' → '{"a": false'
+
+    if (this.#inString) {
+      return ""; // Don't complete anything if we're in an unterminated string
+    }
+
+    const lastChar = output.charAt(output.length - 1);
+
+    // Handle incomplete negative numbers
+    if (lastChar === "-") {
+      return "0"; // Complete to -0
+    }
+
+    // Quick heuristic: only check partial keywords if last char could be part of one
+    if (NULL_KEYWORD_CHARS.has(lastChar)) {
+      if (output.endsWith("nul")) {
+        return "l";
+      } else if (output.endsWith("nu")) {
+        return "ll";
+      } else if (output.endsWith("n")) {
+        return "ull";
+      }
```

Comment:
I wonder if regexes start catching up to `endsWith`'s performance if we do a single regex match vs multiple `endsWith`

Reply:
I gave it a try, and found these results:

- **Small tool calls:** 0.8x **slower** (!) than previous implementation (instead of 1.6x faster)
- **Medium tool calls:** 1.4x faster than previous implementation (instead of 2.6x faster)
- **Large tool calls:** 2.0x faster than previous implementation (instead of 6.3x faster)
- **Character-by-character streaming:** 2.1x faster than previous implementation (instead of 2.2x faster)

So using regexes here is significantly slower than using the `.endsWith()` approach.